File: module-11-vpn-technologies/wg_keygen.py
"""
Module 11 -- generates REAL WireGuard-compatible key pairs. These are not
placeholder strings -- they're genuine X25519 private/public keys, base64
encoded in the exact format `wg genkey` / `wg pubkey` produce, using the
same elliptic-curve primitive (Curve25519) WireGuard itself uses.

You could take this script's output and paste it directly into a real
WireGuard config file and it would work correctly -- the only thing this
course can't do FOR you is install the WireGuard app and run the tunnel
itself (see DECISIONS.md and tutorial.html).

SECURITY NOTE: a WireGuard private key is a secret, equivalent in
sensitivity to an SSH private key or a password. Never commit real
generated keys to a public repository or share them outside your own
devices. Every key this script prints during its own demo run is
discarded the moment the process exits -- generate your OWN keys for any
tunnel you actually intend to use.
"""
import base64

from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey
from cryptography.hazmat.primitives.serialization import (
    Encoding,
    NoEncryption,
    PrivateFormat,
    PublicFormat,
)


def generate_keypair():
    """Returns (private_key_b64, public_key_b64), matching `wg genkey`'s output format exactly."""
    private_key = X25519PrivateKey.generate()
    private_bytes = private_key.private_bytes(
        encoding=Encoding.Raw, format=PrivateFormat.Raw, encryption_algorithm=NoEncryption()
    )
    public_bytes = private_key.public_key().public_bytes(
        encoding=Encoding.Raw, format=PublicFormat.Raw
    )
    return base64.b64encode(private_bytes).decode(), base64.b64encode(public_bytes).decode()


def derive_public_key(private_key_b64: str) -> str:
    """Equivalent to `wg pubkey`: given a private key, derive its public key."""
    private_bytes = base64.b64decode(private_key_b64)
    private_key = X25519PrivateKey.from_private_bytes(private_bytes)
    public_bytes = private_key.public_key().public_bytes(
        encoding=Encoding.Raw, format=PublicFormat.Raw
    )
    return base64.b64encode(public_bytes).decode()


if __name__ == "__main__":
    priv, pub = generate_keypair()
    print(f"PrivateKey = {priv}")
    print(f"PublicKey  = {pub}")

    print("\nVerifying: deriving the public key from the private key independently...")
    rederived = derive_public_key(priv)
    print(f"Re-derived  = {rederived}")
    print(f"Match: {rederived == pub}")

    print(f"\nKey length check: {len(base64.b64decode(priv))} bytes "
          f"(WireGuard/X25519 keys are always exactly 32 raw bytes)")

File: module-11-vpn-technologies/wg_keygen.js
/**
 * Module 11 — JS/Node port of wg_keygen.py. Same real X25519 key
 * generation using Node's built-in `crypto` module (no npm dependency
 * needed -- Node has shipped X25519 support natively since v12).
 *
 * SECURITY NOTE: same as the Python file -- treat generated private keys
 * as real secrets. Never commit them anywhere.
 */
const crypto = require("crypto");

function generateKeypair() {
  const { publicKey, privateKey } = crypto.generateKeyPairSync("x25519");
  const privB64 = Buffer.from(privateKey.export({ format: "jwk" }).d, "base64url").toString("base64");
  const pubB64 = Buffer.from(publicKey.export({ format: "jwk" }).x, "base64url").toString("base64");
  return { priv: privB64, pub: pubB64 };
}

function derivePublicKey(privateKeyB64) {
  const rawPriv = Buffer.from(privateKeyB64, "base64");
  const jwk = {
    kty: "OKP",
    crv: "X25519",
    d: rawPriv.toString("base64url"),
    x: "", // placeholder -- createPrivateKey derives the real public part below
  };
  // Node needs a full JWK with SOME x value to import; derive the real
  // public key afterward via createPublicKey(privateKeyObject), which
  // recomputes it from the private scalar rather than trusting our
  // placeholder.
  const tempPub = crypto.generateKeyPairSync("x25519").publicKey.export({ format: "jwk" }).x;
  jwk.x = tempPub;
  const privateKeyObj = crypto.createPrivateKey({ key: jwk, format: "jwk" });
  const publicKeyObj = crypto.createPublicKey(privateKeyObj);
  const pubJwk = publicKeyObj.export({ format: "jwk" });
  return Buffer.from(pubJwk.x, "base64url").toString("base64");
}

if (require.main === module) {
  const { priv, pub } = generateKeypair();
  console.log(`PrivateKey = ${priv}`);
  console.log(`PublicKey  = ${pub}`);

  console.log("\nVerifying: deriving the public key from the private key independently...");
  const rederived = derivePublicKey(priv);
  console.log(`Re-derived  = ${rederived}`);
  console.log(`Match: ${rederived === pub}`);

  console.log(`\nKey length check: ${Buffer.from(priv, "base64").length} bytes ` +
    `(WireGuard/X25519 keys are always exactly 32 raw bytes)`);
}

module.exports = { generateKeypair, derivePublicKey };
